### ISSUE TITLE
Using a decorator to validate the type of a member doesn't work

### ISSUE THREAD

**Issue by Gryfenfer97** [microsoft/pyright#5388]
**Describe the bug**
When having a decorator inside a class that check a member type, the check isn't applied to the function decorated.

**To Reproduce**
I have made this example:
```python
from typing import Any, Callable, Optional

class Foo:
    def __init__(self):
        self._x: Optional[int] = 5

    @staticmethod
    def _check_x(func: Callable[..., Any]):
        def decorator(self: "Foo"):
            if self._x is None:
                return
            return func(self)
        return decorator

    @_check_x
    def bar(self):
        self._x += 2


foo = Foo()
foo.bar()
```

Here `self._x` should be a int but is still considered as Optional[None]

**VS Code extension**
Pylance v2023.6.30

I chose the label `bug` but this might be an enhancement

**Comment by erictraut** [comment]
This is as designed. Type narrowing is applied only within an execution flow (i.e. within a code flow graph), not across function calls. If you want your code to work with a static type checker, you should adopt a different mechanism for validating the type of a member.
